android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/car/CarCameraController.kt

```diff
@@ -1,60 +1,87 @@
 package com.mapbox.maps.testapp.auto.car
 
+import android.graphics.Rect
 import com.mapbox.geojson.Point
+import com.mapbox.maps.CameraState
 import com.mapbox.maps.EdgeInsets
-import com.mapbox.maps.MapSurface
 import com.mapbox.maps.dsl.cameraOptions
-import com.mapbox.maps.extension.androidauto.OnMapScrollListener
+import com.mapbox.maps.extension.androidauto.MapboxCarMapObserver
+import com.mapbox.maps.extension.androidauto.MapboxCarMapSurface
 import com.mapbox.maps.plugin.animation.camera
 import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener
+import com.mapbox.maps.plugin.locationcomponent.location
 
 /**
  * Controller class to handle map camera changes.
  */
-class CarCameraController : OnIndicatorPositionChangedListener, OnMapScrollListener {
+class CarCameraController : MapboxCarMapObserver {
 
   private var lastGpsLocation: Point = HELSINKI
+  private var previousCameraState: CameraState? = null
   private var isTrackingPuck = true
 
-  private lateinit var surface: MapSurface
+  private var surface: MapboxCarMapSurface? = null
   private var insets: EdgeInsets = EdgeInsets(0.0, 0.0, 0.0, 0.0)
 
+  private val changePositionListener = OnIndicatorPositionChangedListener { point ->
+    lastGpsLocation = point
+    if (isTrackingPuck) {
+      surface?.mapSurface?.getMapboxMap()?.setCamera(
+        cameraOptions {
+          center(point)
+          padding(insets)
+        }
+      )
+    }
+  }
```

Comment:
we can use the new viewport plugin to handle car camera, it also could be addressed in follow up PR.

Reply:
Do you have a link to an example? I could play with this some more

Reply:
yes, it's still in beta, you can checkout the viewport plugin at https://github.com/mapbox/mapbox-maps-android/tree/main/plugin-viewport, and there's 2 examples at https://github.com/mapbox/mapbox-maps-android/tree/main/app/src/main/java/com/mapbox/maps/testapp/examples/viewport

Reply:
going to let this one be follow up

Reply:
I tried it, but this doesn't work for dark/light mode switching 🤔 

``` kotlin
@OptIn(MapboxExperimental::class)
class CarMapShowcase : MapboxCarMapObserver {

  private var mapboxCarMapSurface: MapboxCarMapSurface? = null

  override fun onAttached(mapboxCarMapSurface: MapboxCarMapSurface) {
    loadMapStyle(mapboxCarMapSurface.carContext)
  }

  fun loadMapStyle(carContext: CarContext) {
    // https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#paint-sky-sky-atmosphere-sun
    // Position of the sun center [a azimuthal angle, p polar angle].
    // Azimuthal is degrees from north, where 0.0 is north.
    // Polar is degrees from overhead, where 0.0 is overhead.
    val sunDirection = if (carContext.isDarkMode) listOf(-50.0, 90.2) else listOf(0.0, 0.0)
    val styleUri = if (carContext.isDarkMode) Style.TRAFFIC_NIGHT else Style.TRAFFIC_DAY

    mapboxCarMapSurface?.mapSurface?.getMapboxMap()?.loadStyle(
      style(styleUri) {
        +rasterDemSource(DEM_SOURCE) {
          url(TERRAIN_URL_TILE_RESOURCE)
          // 514 specifies padded DEM tile and provides better performance than 512 tiles.
          tileSize(514)
        }
        +terrain(DEM_SOURCE)
        +skyLayer(SKY_LAYER) {
          skyType(SkyType.ATMOSPHERE)
          skyAtmosphereSun(sunDirection)
        }
      }
    )
  }

  companion object {
    private const val SKY_LAYER = "sky"
    private const val DEM_SOURCE = "mapbox-dem"
    private const val TERRAIN_URL_TILE_RESOURCE = "mapbox://mapbox.mapbox-terrain-dem-v1"
  }
}
```

Reply:
oh found it. never assigned `mapboxCarMapSurface`